ui/app/src/components/PublishOnDemandWidget/PublishOnDemandWidget.tsx

```diff
@@ -348,110 +389,150 @@ export function PublishOnDemandWidget(props: PublishOnDemandWidgetProps) {
     }
   };
 
+  const publishDialogSuccess = 'publishDialogSuccess';
+  const publishDialogClosed = 'publishDialogClosed';
+  const onInitialPublish = () => {
+    dispatch(
+      showPublishDialog({
+        items: [initialPublishItem],
+        onSuccess: batchActions([closePublishDialog(), dispatchDOMEvent({ id: publishDialogSuccess })]),
+        onClosed: dispatchDOMEvent({ id: publishDialogClosed })
+      })
+    );
+
+    let unsubscribe, cancelUnsubscribe;
+    unsubscribe = createCustomDocumentEventListener(publishDialogSuccess, () => {
+      setHasInitialPublish(true);
+      cancelUnsubscribe();
+    });
+    cancelUnsubscribe = createCustomDocumentEventListener(publishDialogClosed, () => {
+      unsubscribe();
+    });
```

Comment:
Only one custom event callback is needed. Simply add parameters to the detail so you can add conditional logic to do what you need to depending on the case. Unsubscribing is done automatically by the `createCustomDocumentEventListener` util.